### PR TITLE
kpatch: retry kpatch load on failure

### PR DESCRIPTION
**Pull request by jpoimboe**
If a kpatch load fails, try again, up to 5 times.

It will usually only fail if the activeness safety check failed, in
which case trying again might work.  Unfortunately, insmod doesn't
return the error reported by the module, so we can't distinguish
activeness safety errors from other types of errors.

### CODE REVIEW

File: kpatch/kpatch [hunk 1]
```diff
@@ -163,7 +163,16 @@ load_module () {
...
 	echo "loading patch module: $1"
-	insmod "$1" "$2"
+	for i in `seq 5`; do
+		if [[ $i != 1 ]]; then
+			echo "retrying..."
+			sleep 2
+		fi
+		insmod "$1" "$2" && break
+	done
+	[[ $i = 5 ]] && die "failed to load module $modname"
```
Comment (flaming-toast) [inline kpatch/kpatch]: If `insmod` happens to succeed on the 5th try (i = 5), then we will `die` here. Maybe set a fail flag within the loop (at i=5, if insmod fails, then set fail=1...) and then check it outside of the for loop?

File: kpatch/kpatch [hunk 2]
```diff
@@ -163,7 +163,16 @@ load_module () {
 	fi
 
 	echo "loading patch module: $1"
-	insmod "$1" "$2"
+	for i in `seq 5`; do
+		if [[ $i != 1 ]]; then
+			echo "retrying..."
+			sleep 2
+		fi
+		insmod "$1" "$2" && break
```
Comment (flaming-toast) [inline kpatch/kpatch]: Do we still need to call `insmod` with `$2`? Looks like nothing in the script sets it anymore. It used to be used to pass the `replace` param, but IIRC the `kpatch replace` option was deprecated in commit 8e8de4718d8

File: kpatch/kpatch (hunk 1 above)
Comment (joe-lawrence) [inline kpatch/kpatch]: Alternatively move the conditional check at the bottom of the loop.  Any successful ```insmod``` will break the loop and skip the check...  Or do something like: ```insmod "$1" "$2" && loaded=1 && break``` to employ a success flag without having duplicate "5" constants laying about.

Comment (flaming-toast) [inline kpatch/kpatch]: @joe-lawrence Ah, I like your idea better than mine :-) (ditching the conditional and moving the sleep and echo after the insmod)

File: kpatch/kpatch (hunk 2 above)
Comment (jpoimboe) [inline kpatch/kpatch]: Oops, yeah.  This is some ugly code.  These functions should really be using local variables to deal with the function arguments.

File: kpatch/kpatch (hunk 1 above)
Comment (jpoimboe) [inline kpatch/kpatch]: Will fix.

**Comment by jpoimboe** [comment]
Updated.

- rewrote loop based on all the comments
- cleaned up load_module() to use local variables
- removed unused 2nd argument to insmod
- added specific checking for insmod error, retry only makes sense for EBUSY.

File: kpatch/kpatch [hunk 3]
```diff
@@ -165,7 +165,27 @@ load_module () {
 	fi
 
 	echo "loading patch module: $module"
-	insmod "$module"
+	local i=0
+	while true; do
+		out=$(insmod "$module" 2>&1)
+		[[ -z $out ]] && break
+		echo $out 1>&2
+		[[ ! $out =~ "Device or resource busy" ]] &&
```
Comment (joe-lawrence) [inline kpatch/kpatch]: I have no i18n experience, but is it assured that insmod output will be in English?  (Too bad it doesn't propagate out the error code as insmod exit status.)

File: kpatch/kpatch [hunk 4]
```diff
@@ -165,7 +165,27 @@ load_module () {
...
+			die "failed to load module $module"
+
+		# "Device or resource busy" means the activeness safety check
+		# failed.  Retry in a few seconds.
+		i=$((i+1))
+		if [[ $i = 5 ]]; then
```
Comment (joe-lawrence) [inline kpatch/kpatch]: I always get tripped up with shell scripts... is ```=``` legit here, or do you need ```-eq``` for numerical test?

File: kpatch/kpatch (hunk 3 above)
Comment (flaming-toast) [inline kpatch/kpatch]: @joe-lawrence That's an interesting question. I looked at the insmod sources from kmod and old module-init-tools sources (superceded by kmod) and they both have always used `strerror` to print the message, *not* `strerror_l`,  which would print the locale-dependent/translated version of the error message. So I think the message will always be in English.

(For the curious, the translated glibc error messages used by `strerror_l` and others can be found here: https://sourceware.org/git/?p=glibc.git;a=tree;f=po;h=b6fb91b5b5b9730f24a0960bf0b0db766fc54c17;hb=HEAD)

File: kpatch/kpatch (hunk 4 above)
Comment (jpoimboe) [inline kpatch/kpatch]: `=` is legit because all variables are really strings in bash.  So it's testing for the string "5".  `-eq` would have worked too.

Comment (joe-lawrence) [inline kpatch/kpatch]: Okay, like I said, it just gets confusing, especially with ```-gt``` vs ```>```.  I did test this in a shell and you're right, the string comparison will match.

**Pull request merged**